packages/babel-helper-split-export-declaration/src/index.js

```diff
@@ -0,0 +1,65 @@
+import * as t from "@babel/types";
+
+const splitExportDeclaration = exportDeclaration => {
+  if (!exportDeclaration.isExportDeclaration()) {
+    throw new Error("Only export declarations can be splitted.");
+  }
+
+  // build specifiers that point back to this export declaration
+  const isDefault = exportDeclaration.isExportDefaultDeclaration();
+  const declaration = exportDeclaration.get("declaration");
+
+  if (isDefault) {
+    if (declaration.isIdentifier()) {
+      throw new Error("It doesn't make sense to split exported identifier.");
+    }
+    const standaloneDeclaration =
+      declaration.isFunctionDeclaration() || declaration.isClassDeclaration();
+
+    let id = declaration.node.id;
+    const liveBinding = !!id;
+
+    if (!id) {
+      id = declaration.scope.generateUidIdentifier("default");
+      if (standaloneDeclaration) {
+        declaration.node.id = t.cloneNode(id);
+      }
+    }
+
+    const updatedDeclaration = standaloneDeclaration
+      ? declaration
+      : t.variableDeclaration("var", [
+          t.variableDeclarator(t.cloneNode(id), declaration.node),
+        ]);
+
+    const updatedExportDeclaration = liveBinding
+      ? t.exportNamedDeclaration(null, [
+          t.exportSpecifier(t.cloneNode(id), t.identifier("default")),
+        ])
+      : t.exportDefaultDeclaration(t.cloneNode(id));
```

Comment:
@loganfsmyth  [commented](https://github.com/babel/babel/pull/6963#discussion_r164818661) on this:
> Even if it doesn't have an ID, cyclic dependencies can still rely on it being hoisted. It's never safe to do export default foo; in this case.

Reply:
I'm not sure when that could have an impact? is it about smth like this?
```js
export default (() => console.log('side effect'))()
```

Reply:
It might be about something like https://github.com/rollup/rollup/issues/1784

Reply:
```

Reply:
```
// index.js
import "./a";

export default function(){
    return 4;
}


// a.js
import getFour from "./index";

getFour();
```

Reply:
ok, from what I understand this can affect anonymous functions and classes, right? import/exports are hoisted so with the given example we end up with
```
import "./a";

function _default() {
  return 4;
}

export default _default;
```
which is later transformed to by cjs transform
```
"use strict";

exports.__esModule = true;
exports.default = _default;

require("./a");

function _default() {
  return 4;
}
```

everything seems to be ok, `a.js` should be able to import/require it - as those anonymous default exports are splitted into an identifier + named function/class declaration

I suspect that there might be problems with non-hoisted by default things being splitted, right? Like in
```
// index.js
import "./a";

export default 4;

// a.js
import four from "./index";
console.log(four)
```

this would actually end up as
```js
// index.js
exports.__esModule = true;
exports.default = void 0;

require("./a");

var _default = 4;
exports.default = _default;

// a.js
var _index = _interopRequireDefault(require("./index"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

console.log(_index.default);
```

and break - I'll work on fixing this later today. Thanks for catching this!

Reply:
I can't tell from your comment if I was actually clear enough, so just to clarify:

Reply:
I can't tell from your comment if I was actually clear enough, so just to clarify:

The issue is that
```
import "./a";

function _default() {
  return 4;
}

export default _default;
```
exports the default as an expression, meaning the `export default _default` line must have been evaluated before the imported default function will be available, whereas
```
export { _default as default };
```
explicitly re-exports the `_default` binding, which means it is exported up front before the imports are processed.

Mostly It doesn't seem like there's anything to gain from this optimization. 

```
export { _default as default };
```
is universally more consistent than
```
export default _default;
```
so essentially all we are gaining is a couple bytes.

Reply:
It seems that we are still not on the same page, probably because I don't know spec that well as you do. I've tried to look into it right now to clarify things, but man, this thing is still too hard for me to comprehend.

I realize that I do not understand how import/export hoisting works **exactly**.

Let's look on the given example of:
```js
import "./a";

function _default() {
  return 4;
}

export default _default;
```

gets transpiled by the **master** branch to 
```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("./a");

function _default() {
  return 4;
}

var _default2 = _default;
exports.default = _default2;
```

and on my PR to 
```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

require("./a");

function _default() {
  return 4;
}
```

Which is actually incorrect, do I get it correct? Because we are exporting an expression it shouldnt be hoisted right away and assigned to `exports.default`.

It is correct by accident in tests etc because this is an output of splitting the:
```js
export default function () {
  return 4;
}
```
which should get hoisted and all, but after we do the split the intermediate output is wrong (or rather the intermediate output is wrongfully transformed into "correct" output).

I believe transforming into `export { _default as default };` is correct approach IF `_default` should get hoisted. If we split `export default 4` I'd argue that the current way (transforming into `export default _default`) is more correct, because it doesn't "allow" for changing value of `_default` binding afterwards and thus changing the exported value. Such export is not live.

Reply:
> Which is actually incorrect, do I get it correct? Because we are exporting an expression it shouldnt be hoisted right away and assigned to exports.default.

Yep, that's right.

> I'd argue that the current way (transforming into export default _default) is more correct, because it doesn't "allow" for changing value of _default binding afterwards and thus changing the exported value. Such export is not live.

It seems like an overoptimization to me. If some other plugin is mutating bindings locally, they already run the risk of mutating a live binding of something they weren't supposed to if the user wrote their code without ever reassigning an exported variable. Also it seems counter to what I mentioned in my other comment for what the overall goal of this module should be. I view the goal of splitting specifically to separate an export into a standard JS statement with no export info, and an export statement without runtime behavior. By splitting it into `export default foo;`, we've split one statement with an expression into two separate statements with expressions.